Add `typeParams` to `ArrowTypeInfo` (#1504)

ktonga opened the pull request:
I don't totally agree on some of the coding decisions but I tried to follow the style to not introduce unrelated changes. Also, I know there are a lot of tickets for improving the codebase in general, I hope to be able to help with that!

And obviously, @fommil,  I want my ENSIME sticker(s?) :-)

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/ensime/ensime-server?pullRequest=1539) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/ensime/ensime-server?pullRequest=1539) before we can accept your contribution.<br/>

fommil commented:
Cool. Only a quick look but some comments: use List instead of Iterable. Don't add the companion, we try to keep these files as pure ADT with no code (have to break the rule where the JVM forces us for toString etc).

ktonga commented:
Ok, I used iterable just to follow the other fields. I can change it.
Ok with the companion, will remove it. But are you ok with having the shortcuts somewhere? it makes no sense to repeat the same four parameters over and over again in creation and pattern matching.

fommil commented:
Yeah that's fine, if our domain model is repetitive it probably hints that the model is broken.

fommil commented:
Changing Iterable to List would be nice :-P

ktonga commented:
It's broken in a sense that they are not truly ADTs 'cause fields are unnecessarily defined in a higher trait, when they could be in the concrete case class which needs the field (it seems like old fashioned OOP habit). I'd like to participate if you are planning on reworking the API/Model.

I've already changed to `List` the field I've introduced. You mean I should change all of them?

ktonga commented:
Done. I've moved the companion to the model package, near the Builders.

fommil commented:
yeah I agree that it's not a true ADT. Same is true of the `domain.scala` file in the indexer. There is definitely some work in both packages to move to a true ADT with typeclasses for all the functionality, but I've just always had higher priority things to work on. If you're looking for general clean-ups that you could get involved in, it would be very much welcomed... indeed I'd even consider moving the FQN domain objects into `api` (which would require a change of license but I think everybody will agree).

fommil commented on api/src/test/scala/org/ensime/api/EnsimeTestData.scala:
```diff
@@ -14,7 +14,7 @@ trait EnsimeTestData {
   }
 
-  val typeInfo = new BasicTypeInfo("type1", DeclaredAs.Method, "FOO.type1", List(), List(), None)
+  val typeInfo = BasicTypeInfo("type1", DeclaredAs.Method, "FOO.type1", Nil, Nil, None, Nil)
```
we use `new` in a whole load of places, it's because of the path that the code has taken... it wasn't always as clean an ADT as it is today and used to be combined with the modelbuilder file.

fommil commented:
LGTM, the failing tests look like transients. Travis has been acting up, not sure what's wrong with drone.

fommil commented:
you've found some nice code clean-ups here as you've gone along. It's well appreciated.

fommil commented:
emacs tests failed again, but looks unrelated to your changes. @chipsenkbeil how is your machine looking?

chipsenkbeil commented:
@fommil, just SSH'd in. Looks fine. I can see something running on it. Compiling fine so far.

ktonga commented:
Yeap, dunno what's wrong with Drone and Travis. FTR I've run `test` and `it:test` locally and they succeeded, including the new cases I've introduced.

Looking forward to keep contributing. Ensime is an amazing piece of Software. I'll give it a try to replace Akka with FS2 in SearchService, let's see how it goes.

fommil commented:
Monix is looking like a viable alternative too ;-)

ktonga commented:
I know, I was following the ticket closely. But my heart is with FS2, it brings such an elegant programming model and they are working hard to improve performance.

ktonga commented:
You need anything else from me for this PR?

fommil commented:
I'll have a look over the next few days.

fommil commented:
I still think these are transient errors. Travis looks like it picked up a bad cache at some point, so I'll clear that out and restart. No idea what's going on with drone.

fommil commented:
on the SearchService, I think it's blocked by #1481 ... but if you're keen on something streamy then looking into replacing akka-stream websockets with something else would be really good. Although we'd prefer Java libraries instead of anything too fancy, to make it easier to support scala 2.12.

fommil commented:
Try pushing a commit. I'm starting to think this isn't a transient on drone.

ktonga commented:
I've squashed. The sbt build in drone finishes ok, it's a timeout in the ensime-emacs tests I think.

ktonga commented:
It failed again with same error, it can't update Ensime server. I don't know what that means. Not an emacs expert precisely.

chipsenkbeil commented:
@fommil, looks like this problem has also happened on your machine several days ago; so, I don't think this is a machine-specific problem. Seems to be plaguing the 2.11.8 build.

fommil commented:
The emacs tests are passing for other prs so it makes me wonder.

fommil commented:
I'll investigate locally to see if I spot anything obvious

fommil commented:
btw, you raised the PR against 1.0, not 2.0. Can you please re-raise against the dev branch? 1.0 is in critical bugfixes only mode. This implies that we have a broken CI for the 1.0 branch that I'll need to fix.
